File: frontend/src/api/assetLedger.ts
import apiClient from './client'

export interface MetaOption { value: string; label: string }

export interface LedgerMeta {
  entry_types: MetaOption[]
  depreciation_methods: MetaOption[]
  // Served rather than hardcoded here so the two cannot disagree about how
  // long a lift lasts.
  default_useful_life_years: Record<string, number>
}

export interface PeriodRow {
  year: number
  opening_book_value: string
  depreciation: string
  accumulated: string
  closing_book_value: string
  // Non-zero where an improvement or impairment landed in this year, which is
  // why the curve bends.
  basis_change: string
  note: string
}

export interface Depreciation {
  method: string
  cost: string
  salvage_value: string
  useful_life_years: string
  in_service_date: string | null
  as_of: string
  depreciable_amount: string
  accumulated_depreciation: string
  net_book_value: string
  annual_depreciation: string
  monthly_depreciation: string
  months_elapsed: number
  months_remaining: number
  percent_depreciated: number
  is_fully_depreciated: boolean
  schedule: PeriodRow[]
  // Present when a figure could not be computed. Distinguishes "no cost on
  // record" from "worth nothing", which look identical as a zero.
  message: string | null
}

export interface ServiceSpend {
  completed_work_orders: number
  corrective_count: number
  preventive_count: number
  total_service_cost: string
  total_labour_hours: string
}

export interface TimelineEvent {
  kind: string
  occurred_on: string | null
  title: string
  detail: string | null
  amount: string | null
  reference: string | null
  source: string | null
  source_id: number | null
  outcome: string | null
}

export interface LedgerEntry {
  id: number
  facility_id: number
  equipment_id: number
  entry_type: string
  effective_date: string
  description: string
  amount: string | null
  proceeds: string | null
  gain_loss: string | null
  from_location_id: number | null
  to_location_id: number | null
  extends_useful_life_years: string | null
  reference: string | null
  vendor_id: number | null
  work_order_id: number | null
  reverses_entry_id: number | null
  is_reversed: boolean
  notes: string | null
  created_at: string
}

export interface AssetLedgerSummary {
  equipment_id: number
  asset_tag: string | null
  in_service_date: string | null
  disposed_on: string | null
  age_months: number | null
  depreciation: Depreciation
  service: ServiceSpend
  // Repair cost approaching replacement cost says something the depreciation
  // schedule cannot.
  service_cost_as_percent_of_cost: number | null
  ledger_entry_count: number
}

export interface AssetLedger {
  summary: AssetLedgerSummary
  timeline: TimelineEvent[]
  entries: LedgerEntry[]
}

export interface DisciplineValuation {
  asset_count: number
  cost: string
  accumulated_depreciation: string
  net_book_value: string
}

export interface FleetValuation {
  as_of: string
  asset_count: number
  total_cost: string
  accumulated_depreciation: string
  net_book_value: string
  // Kit still in service with no book value left is a replacement nobody has
  // budgeted for.
  fully_depreciated_count: number
  by_discipline: Record<string, DisciplineValuation>
}

export const fetchLedgerMeta = async (): Promise<LedgerMeta> => {
  const res = await apiClient.get('/asset-ledger/meta')
  return res.data
}

export const fetchAssetLedger = async (
  equipmentId: number,
  asOf?: string,
): Promise<AssetLedger> => {
  const res = await apiClient.get(`/asset-ledger/equipment/${equipmentId}`, {
    params: asOf ? { as_of: asOf } : undefined,
  })
  return res.data
}

/**
 * The schedule on its own, with optional what-if overrides. Finance asks
 * "what would fifteen years look like instead of twenty" often enough that
 * answering without writing anything is worth two query parameters.
 */
export const fetchDepreciation = async (
  equipmentId: number,
  params: { as_of?: string; method?: string; useful_life_years?: number } = {},
): Promise<Depreciation> => {
  const res = await apiClient.get(`/asset-ledger/equipment/${equipmentId}/depreciation`, { params })
  return res.data
}

export const fetchFleetValuation = async (params: {
  facility_id?: number
  as_of?: string
} = {}): Promise<FleetValuation> => {
  const res = await apiClient.get('/asset-ledger/valuation', { params })
  return res.data
}

export const fetchLedgerEntries = async (params: {
  equipment_id?: number
  facility_id?: number
  entry_type?: string
  since?: string
  limit?: number
}): Promise<{ items: LedgerEntry[]; total: number }> => {
  const res = await apiClient.get('/asset-ledger/entries', { params })
  return res.data
}

export const createLedgerEntry = async (payload: {
  equipment_id: number
  entry_type: string
  effective_date: string
  description: string
  amount?: number | null
  proceeds?: number | null
  from_location_id?: number | null
  to_location_id?: number | null
  extends_useful_life_years?: number | null
  reference?: string | null
  vendor_id?: number | null
  work_order_id?: number | null
  notes?: string | null
}): Promise<LedgerEntry> => {
  const res = await apiClient.post('/asset-ledger/entries', payload)
  return res.data
}

/**
 * Corrections are posted, not edited. A financial row somebody can see was
 * rewritten is a row an auditor cannot rely on.
 */
export const reverseLedgerEntry = async (
  entryId: number,
  reason: string,
): Promise<LedgerEntry> => {
  const res = await apiClient.post(`/asset-ledger/entries/${entryId}/reverse`, { reason })
  return res.data
}

File: backend/app/schemas/asset_ledger.py
from typing import Any, Dict, List, Optional
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field, field_validator

from app.models.asset_ledger import DEPRECIATION_METHODS, LEDGER_ENTRY_TYPES


class LedgerEntryCreate(BaseModel):
    equipment_id: int
    entry_type: str
    # The date the event happened, not the date it was recorded. An improvement
    # completed in March and entered in June depreciates from March.
    effective_date: date
    description: str = Field(min_length=1, max_length=255)
    amount: Optional[Decimal] = None
    proceeds: Optional[Decimal] = None
    from_location_id: Optional[int] = None
    to_location_id: Optional[int] = None
    to_facility_id: Optional[int] = None
    extends_useful_life_years: Optional[Decimal] = None
    reference: Optional[str] = None
    vendor_id: Optional[int] = None
    work_order_id: Optional[int] = None
    notes: Optional[str] = None

    @field_validator("entry_type")
    @classmethod
    def _known_type(cls, value: str) -> str:
        if value not in LEDGER_ENTRY_TYPES:
            raise ValueError(
                f"Unknown entry type '{value}'. Allowed: {', '.join(LEDGER_ENTRY_TYPES)}"
            )
        return value


class LedgerEntryReverse(BaseModel):
    """Corrections are posted, not edited.

    A financial row an auditor can see was rewritten is a row they cannot rely
    on, so the original stays and a reversing entry is posted against it.
    """

    reason: str = Field(min_length=1, max_length=255)


class LedgerEntry(BaseModel):
    id: int
    facility_id: int
    equipment_id: int
    entry_type: str
    effective_date: date
    description: str
    amount: Optional[Decimal] = None
    proceeds: Optional[Decimal] = None
    gain_loss: Optional[Decimal] = None
    from_location_id: Optional[int] = None
    to_location_id: Optional[int] = None
    extends_useful_life_years: Optional[Decimal] = None
    reference: Optional[str] = None
    vendor_id: Optional[int] = None
    work_order_id: Optional[int] = None
    reverses_entry_id: Optional[int] = None
    is_reversed: bool
    notes: Optional[str] = None
    created_by_id: Optional[int] = None
    created_at: datetime

    class Config:
        from_attributes = True


class LedgerEntryListResponse(BaseModel):
    items: List[LedgerEntry]
    total: int


class PeriodRow(BaseModel):
    year: int
    opening_book_value: Decimal
    depreciation: Decimal
    accumulated: Decimal
    closing_book_value: Decimal
    # Non-zero where an improvement or impairment landed in this year, so a
    # reader can see why the curve bends.
    basis_change: Decimal = Decimal("0")
    note: str = ""


class Depreciation(BaseModel):
    method: str
    cost: Decimal
    salvage_value: Decimal
    useful_life_years: Decimal
    in_service_date: Optional[date] = None
    as_of: date
    depreciable_amount: Decimal
    accumulated_depreciation: Decimal
    net_book_value: Decimal
    annual_depreciation: Decimal
    monthly_depreciation: Decimal
    months_elapsed: int
    months_remaining: int
    percent_depreciated: float
    is_fully_depreciated: bool
    schedule: List[PeriodRow] = []
    # Says why a figure is missing rather than returning a silent zero, which
    # reads identically to an asset that cost nothing.
    message: Optional[str] = None


class ServiceSpend(BaseModel):
    completed_work_orders: int
    corrective_count: int
    preventive_count: int
    total_service_cost: Decimal
    total_labour_hours: Decimal


class TimelineEvent(BaseModel):
    kind: str
    occurred_on: Optional[date] = None
    title: str
    detail: Optional[str] = None
    amount: Optional[Decimal] = None
    reference: Optional[str] = None
    source: Optional[str] = None
    source_id: Optional[int] = None
    outcome: Optional[str] = None


class AssetLedgerSummary(BaseModel):
    equipment_id: int
    asset_tag: Optional[str] = None
    in_service_date: Optional[date] = None
    disposed_on: Optional[date] = None
    age_months: Optional[int] = None
    depreciation: Depreciation
    service: ServiceSpend
    # An asset whose cumulative repair cost approaches its original cost is
    # telling you something the depreciation schedule cannot.
    service_cost_as_percent_of_cost: Optional[float] = None
    ledger_entry_count: int


class AssetLedger(BaseModel):
    summary: AssetLedgerSummary
    timeline: List[TimelineEvent]
    entries: List[LedgerEntry]


class DisciplineValuation(BaseModel):
    asset_count: int
    cost: Decimal
    accumulated_depreciation: Decimal
    net_book_value: Decimal


class FleetValuation(BaseModel):
    as_of: date
    asset_count: int
    total_cost: Decimal
    accumulated_depreciation: Decimal
    net_book_value: Decimal
    # Kit still in service with no book value left is kit whose replacement
    # nobody has budgeted for.
    fully_depreciated_count: int
    by_discipline: Dict[str, DisciplineValuation]


class LedgerMeta(BaseModel):
    entry_types: List[Dict[str, str]]
    depreciation_methods: List[Dict[str, str]]
    default_useful_life_years: Dict[str, int]
